### PR TITLE
Don't have the PA params specify multiple helpers

### PR DESCRIPTION
**Pull request by cjpatton**
Partially addresses #68. (We also need to decide if/how to allow a particular PA protocol to use multiple helpers for privacy.)

Right now, multiple helpers are used for redundancy in case one of the
helpers dropped out. This removes this mechanism from the core protocol.

### CODE REVIEW

**Comment by cjpatton** [comment]
Squashed and rebased.

File: draft-pda-protocol.md [hunk 1]
```diff
@@ -504,10 +499,11 @@ The leader's response to malformed requests is specified in
...
 
-For each URL `[helper]` in `PAUploadStartResp.helper_urls`, the client sends a
-GET request to `[helper]/key_config`. The helper responds with status 200 and an
-`HpkeConfig` message. Next, the client collects the set of helpers it will
-upload shares to. It ignores a helper if:
+Let `[helper]` denote the helper URL encoded by
+`PAUploadStartResp.param.helper_url`.  When a client sends a GET request to
+`[helper]/key_config`, the helper responds with status 200 and an `HpkeConfig`
+message. The client aborts and alerts the leader with "unsupported helper" if
+any of the following happen:
```
Comment (chris-wood) [inline draft-pda-protocol.md:506]:
```suggestion
message. The client aborts if any of the following happen:
```

**Review by chris-wood — APPROVED** [review]
LGTM. Multiple helpers "for privacy" seems like it's inherently a per-protocol thing (Prio supports it, Hits does not). If and when we decide to address that, should we move this to the protocol-specific slot in PAParam?

**Comment by cjpatton** [comment]
> LGTM. Multiple helpers "for privacy" seems like it's inherently a per-protocol thing (Prio supports it, Hits does not). If and when we decide to address that, should we move this to the protocol-specific slot in PAParam?

That's a good idea,

**Review by tgeoghegan — APPROVED** [review]

**Comment by ekr** [comment]
> LGTM. Multiple helpers "for privacy" seems like it's inherently a per-protocol thing (Prio supports it, Hits does not). If and when we decide to address that, should we move this to the protocol-specific slot in PAParam?

Hmm... Do we need to decide. Per our discussion on Wednesday, if the clients learn this out of band then we won't need this field at all because clients will get it separately and helpers don't need to know what the other helpers are, right?

**Comment by chris-wood** [comment]
> Hmm... Do we need to decide. Per our discussion on Wednesday, if the clients learn this out of band then we won't need this field at all because clients will get it separately and helpers don't need to know what the other helpers are, right?

My takeaway is that the parameters must contain all input needed to run the protocol, including the helpers. And all of this is configured out of band. You could punt everything except for the cryptographic properties out of the parameters structure, forcing clients to get public keys out of band as well, which would be fine. It's just different. I think both work just fine, but I lean towards having params specify everything. It just seems simpler to me.

**Pull request merged**